[Arm] Fix generating code with UB in NeonEmitter

momchil-velikov opened the pull request:
When generating `arm_neon.h`, NeonEmitter outputs code that
violates strict aliasing rules (C23 6.5 Expressions https://github.com/llvm/llvm-project/pull/7,
C++23 7.2.1 Value category [basic.lval] https://github.com/llvm/llvm-project/pull/11), for
example

    bfloat16_t __reint = __p0;
    uint32_t __reint1 = (uint32_t)(*(uint16_t *) &__reint) << 16;
    __ret = *(float32_t *) &__reint1;

This patch fixed the offending code by replacing it with
a call to `__builtin_bit_cast`.

CarolineConcatto reviewed (commented):
Hi Momchil,
The patch looks good to me. I believe we can also use bit cast for the neon tuples, like I left in the comments.

momchil-velikov commented on clang/utils/TableGen/NeonEmitter.cpp:
```diff
@@ -1382,14 +1382,15 @@ void Intrinsic::emitBodyAsBuiltinCall() {
-      if (CastToType.isInteger())
-        CastToType.makeSigned();
-      Arg = "(" + CastToType.str() + ")" + Arg;
+    if (CastToType.isVector()) {
```
It's an omission. Also the return value cast, it's either redundant or needs to be a bitcast too. Will fix.

momchil-velikov replied:
This change is postponed for another time.

jthackray reviewed (approved):
LGTM

momchil-velikov committed:
[AArch64] Fix generating a code with UB in NeonEmitter

When generating `arm_neon.h`, NeonEmitter outputs code that
violates strict aliasing rules (C23 6.5 Expressions #7,
C++23 7.2.1 Value category [basic.lval] #11).

This patch fixed the offending code by replacing it with
a call to `__builtin_bit_cast`.

momchil-velikov commented:
I reduced the scope of this PR to just fixing the immediate UB.
